src/python/pants/base/build_environment.py

```diff
@@ -50,23 +54,46 @@ def is_in_container() -> bool:
     )
 
 
-class _GitIinitialized(Enum):
+class _GitInitialized(Enum):
     NO = 0
 
 
-_Git: _GitIinitialized | Git | None = _GitIinitialized.NO
+_Git: _GitInitialized | GitResult = _GitInitialized.NO
 
 
-def get_git() -> Git | None:
+@dataclasses.dataclass(frozen=True)
+class GitResult(EngineAwareReturnType):
```

Comment:
Maybe call this Git binary? And put it in the system_binaries?

Or, maybe even more clear is to rename the original type to GitBinary. And then call this MaybeGetBinary?

Reply:
There are a few considerations here:
1. The wrapper is fallible for two reasons:
    a. `git` might not be present
    b. Pants might not be running inside of a `git` working copy
2. The `Git` class has methods that we want to expose, so a fallible ("maybe"/"optional") wrapper type around it should probably stay simple, so that the user needs to unwrap or raise to get access.
3. The mutable/side-effecting methods of the `Git` class should move to another class... maybe a subclass for use only in tests.
    * It looks like the only tests which use it are `src/python/pants/vcs/changed_integration_test.py` and its own test at `src/python/pants/vcs/git_test.py`. 

> Maybe call this Git binary? And put it in the system_binaries?

Putting it in `system_binaries` would be fine I think, yea. But the `get_git` callsite would need to actually be called as a `@rule` in order to be able to use the `system_binaries` infrastructure. That should be fine though. And the `_GitInitialized` infrastructure could actually be deleted, since the engine would handle memoizing the binary lookup.